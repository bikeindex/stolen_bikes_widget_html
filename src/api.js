const API_ENDPOINT = `https://bikeindex.org/api/v2/bikes_search/stolen?per_page=10&widget_from=${
  window.location.hostname
}`;

const request = async url => {
  const resp = await fetch(url);
  const json = await resp.json();
  return json;
};

/*
  PUBLIC
*/
const fetchStolenNearby = async location => {
  const url = `${API_ENDPOINT}&proximity=${location}&proximity_radius=100`;
  const results = await request(url);
  return results;
};

const fetchStolenSerial = async serialNumber => {
  const url = `${API_ENDPOINT}&serial=${serialNumber}`;
  const results = await request(url);
  return results;
};

export {
  fetchStolenNearby,
  fetchStolenSerial,
};
