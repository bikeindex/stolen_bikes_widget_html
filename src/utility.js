import dayjs from 'dayjs';

const defaultElementId = 'binx-stolen-widget';
const defaultHeight = 400;
const headerHeight = 41;
const cacheExpiry = 10800000;

const formatDate = str => {
  const date = dayjs.unix(str);
  const today = dayjs().startOf('day');
  const yesterday = dayjs().subtract(1, 'day').startOf('day');

  const formattedDate = date.isSame(today, 'day')
    ? 'Today'
    : date.isSame(yesterday, 'day')
      ? 'Yesterday'
      : date.format('ddd MMM DD YYYY');
  return formattedDate;
};

export {
  formatDate,
  defaultHeight,
  headerHeight,
  cacheExpiry,
  defaultElementId,
};
