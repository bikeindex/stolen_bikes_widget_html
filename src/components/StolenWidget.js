import React, { Component } from 'react';
import PropTypes from 'prop-types';
import searchIcon from '../icons/search.svg';
import List from './List';
import NoResults from './NoResults';
import Loading from './Loading';
import { fetchStolenNearby, fetchStolenSerial } from '../api';
import { headerHeight, defaultHeight, cacheExpiry } from '../utility';
import '../styles/stolen-widget.scss';

export default class StolenWidget extends Component {
  state = {
    results: [],
    loading: true,
    searchToken: '',
    serialNumber: '',
    recentStolen: true,
  }

  async componentDidMount() {
    let results;
    const { location, cacheResults = true, recentResults = true } = this.props;

    if (!recentResults) {
      return this.setState({ loading: false, recentStolen: false });
    }

    if (!cacheResults) {
      ({ bikes: results } = await fetchStolenNearby(location));
    } else {
      const cacheKey = `recentStolen-${location}`;
      const cached = JSON.parse(localStorage.getItem(cacheKey));
      const now = new Date().getTime();
      const expires = (now + cacheExpiry);
      const expiredAt = cached ? cached.expires : 0;

      if (now > expiredAt) {
        ({ bikes: results } = await fetchStolenNearby(location));
        const newCache = JSON.stringify({ results, expires });
        localStorage.setItem(cacheKey, newCache);
      } else {
        ({ results } = cached);
      }
    }

    return this.setState({ loading: false, results });
  }

  searchSerial = async () => {
    this.setState({ loading: true });
    const { searchToken: serialNumber } = this.state;
    const { bikes: results } = await fetchStolenSerial(serialNumber);
    this.setState({
      recentStolen: false, loading: false, results, serialNumber,
    });
  };

  onClickSearch = e => {
    e.preventDefault();
    this.searchSerial();
  };

  onSubmitSearch = e => {
    e.preventDefault();
    this.searchSerial();
  };

  onChangeSerial = e => {
    const searchToken = e.target.value;
    this.setState({ searchToken });
  };

  render() {
    const { location, height } = this.props;
    const {
      loading, serialNumber, searchToken, results, recentStolen,
    } = this.state;
    const noResults = !loading && results.length === 0 && (serialNumber || recentStolen);
    const maxHeight = (Number(height) || defaultHeight) - headerHeight;

    return (
      <div id="stolen-widget">
        <form className="topsearcher" onSubmit={this.onSubmitSearch}>
          <input
            type="text"
            value={searchToken}
            onChange={this.onChangeSerial}
            placeholder="Search for a serial number"
            disabled={loading}
          />
          <button className="subm" type="submit" onClick={this.onClickSearch}>
            <img alt="search" src={searchIcon} />
          </button>
        </form>

        <div className="binxcontainer" id="binx_list_container">
          {loading ? (
            <Loading />
          ) : noResults ? (
            <NoResults
              recentStolen={recentStolen}
              serialNumber={serialNumber}
            />
          ) : (
            <List
              results={results}
              recentStolen={recentStolen}
              serialNumber={serialNumber}
              maxHeight={maxHeight}
            />
          )}

          {!loading && recentStolen && !noResults && (
            <div className="widget-info">
              Recent reported stolen bikes
              {location && <span> near <em>{location}</em></span>}
            </div>
          )}
        </div>
      </div>
    );
  }
}

StolenWidget.propTypes = {
  cacheResults: PropTypes.bool,
  recentResults: PropTypes.bool,
  location: PropTypes.string,
  height: PropTypes.number,
};
