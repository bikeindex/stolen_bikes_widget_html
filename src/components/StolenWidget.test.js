import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import StolenWidget from './StolenWidget';
import * as api from '../api';
import { headerHeight } from '../utility';

jest.mock('../api');

describe('<StolenWidget /> shallow rendered', () => {
  it('matches the snapshot', () => {
    const tree = shallow(<StolenWidget />);
    expect(toJson(tree)).toMatchSnapshot();
  });

  it('accepts config props', () => {
    const wrap = shallow(
      <StolenWidget
        cacheResults
        recentResults={false}
        location="Portland, OR"
        height={400}
      />,
    );
    const { props } = wrap.instance();
    const expected = {
      cacheResults: true,
      recentResults: false,
      location: 'Portland, OR',
      height: 400,
    };
    expect(props).toEqual(expected);
  });

  it('has a disabled search input', () => {
    const wrap = shallow(<StolenWidget />);
    expect(wrap.find('.topsearcher input').is('[disabled]')).toBe(true);
  });

  it('on search input changes searchToken', () => {
    const wrap = shallow(<StolenWidget />);
    const input = wrap.find('input');
    input.simulate('change', { target: { value: 'something' } });
    expect(wrap.state('searchToken')).toEqual('something');
  });

  it('calls searchSerial when the search button is clicked', () => {
    const wrap = shallow(<StolenWidget />);
    const event = Object.assign(jest.fn(), { preventDefault: () => {} });
    const instance = wrap.instance();

    jest.spyOn(instance, 'searchSerial');
    wrap.find('button').simulate('click', event);
    expect(instance.searchSerial).toHaveBeenCalled();
  });
});

describe('<StolenWidget /> mount rendering', () => {
  it('matches the snapshot', () => {
    const tree = mount(<StolenWidget />);
    expect(toJson(tree)).toMatchSnapshot();
  });

  it('has nearby stolen bikes when recentResults is enabled', () => {
    const wrap = mount(<StolenWidget recentResults />);
    expect(wrap.state('results')).not.toHaveLength(0);
  });

  it('does not have nearby stolen bikes when recentResults is disabled', () => {
    const wrap = mount(<StolenWidget recentResults={false} />);
    expect(wrap.state('results')).toHaveLength(0);
  });

  it('uses the location option when searching', () => {
    jest.spyOn(api, 'fetchStolenNearby');
    mount(<StolenWidget location="Portland, OR" />);
    expect(api.fetchStolenNearby).toHaveBeenCalledWith('Portland, OR');
  });

  it('searches for a matching serialNumber', () => {
    const searchToken = '000111222aaa';
    jest.spyOn(api, 'fetchStolenSerial');
    const wrap = mount(<StolenWidget recentResults={false} />);
    wrap.setState({ searchToken });
    wrap.find('button').simulate('click');
    expect(api.fetchStolenSerial).toHaveBeenCalledWith(searchToken);
  });

  it('caches results in localStorage', () => {
    mount(<StolenWidget cacheResults location="Portland, OR" />);
    expect(localStorage.setItem).toHaveBeenCalled();
  });

  it('sets the maxHeight of the list', () => {
    const height = 200;
    const wrap = mount(<StolenWidget height={height} />);
    const expected = height - headerHeight;
    expect(wrap.find('List').prop('maxHeight')).toEqual(expected);
  });
});
