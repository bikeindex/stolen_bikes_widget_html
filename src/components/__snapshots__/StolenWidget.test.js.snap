// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StolenWidget /> mount rendering matches the snapshot 1`] = `
<StolenWidget>
  <div
    id="stolen-widget"
  >
    <form
      className="topsearcher"
      onSubmit={[Function]}
    >
      <input
        disabled={false}
        onChange={[Function]}
        placeholder="Search for a serial number"
        type="text"
        value=""
      />
      <button
        className="subm"
        onClick={[Function]}
        type="submit"
      >
        <img
          alt="search"
          src="test-file-stub"
        />
      </button>
    </form>
    <div
      className="binxcontainer"
      id="binx_list_container"
    >
      <List
        maxHeight={359}
        recentStolen={true}
        results={
          Array [
            Object {
              "date_stolen": 1553295600,
              "frame_colors": Array [
                "Red",
              ],
              "frame_material": null,
              "frame_model": "Strada",
              "handlebar_type": "Drop",
              "id": 596985,
              "is_stock_img": false,
              "large_img": "https://files.bikeindex.org/uploads/Pu/149821/large_20190320_212118.jpg",
              "manufacturer_name": "Novara",
              "serial": "U161K02722",
              "stolen": true,
              "stolen_location": "Portland,OR,97202",
              "thumb": "https://files.bikeindex.org/uploads/Pu/149821/small_20190320_212118.jpg",
              "title": "2017 Novara Strada",
              "year": 2016,
            },
          ]
        }
        serialNumber=""
      >
        <ul
          style={
            Object {
              "maxHeight": 359,
            }
          }
        >
          <li
            className="stolen"
            key="596985"
          >
            <a
              className="stolen-thumb"
              href="https://bikeindex.org/bikes/596985"
              rel="noopener noreferrer"
              target="_blank"
            >
              <img
                alt="thumb"
                src="https://files.bikeindex.org/uploads/Pu/149821/small_20190320_212118.jpg"
              />
            </a>
            <h4>
              <a
                href="https://bikeindex.org/bikes/596985"
                rel="noopener noreferrer"
                target="_blank"
              >
                2017 Novara Strada
              </a>
            </h4>
            <p>
              <span
                className="stolen-color"
              >
                Stolen
              </span>
              <span>
                 from 
                Portland,OR,97202
                 â€”
              </span>
              <span
                className="date-stolen"
              >
                 
                Fri Mar 22 2019
              </span>
            </p>
            <p>
              Serial: 
              <span
                className="serial-text"
              >
                U161K02722
              </span>
            </p>
          </li>
        </ul>
      </List>
      <div
        className="widget-info"
      >
        Recent reported stolen bikes
      </div>
    </div>
  </div>
</StolenWidget>
`;

exports[`<StolenWidget /> shallow rendered matches the snapshot 1`] = `
<div
  id="stolen-widget"
>
  <form
    className="topsearcher"
    onSubmit={[Function]}
  >
    <input
      disabled={true}
      onChange={[Function]}
      placeholder="Search for a serial number"
      type="text"
      value=""
    />
    <button
      className="subm"
      onClick={[Function]}
      type="submit"
    >
      <img
        alt="search"
        src="test-file-stub"
      />
    </button>
  </form>
  <div
    className="binxcontainer"
    id="binx_list_container"
  >
    <Loading />
  </div>
</div>
`;
